Open Judge System/Services/OJS.Services/Data/ParticipantScores/ParticipantScoresDataService.cs

```diff
@@ -119,31 +131,41 @@ public void Delete(IEnumerable<ParticipantScore> participantScoresEnumerable)
             this.participantScores.SaveChanges();
         }
 
-        public void AddBySubmissionByUsernameAndIsOfficial(Submission submission, string username, bool isOfficial)
+        public void AddBySubmissionByUsernameAndIsOfficial(
+            Submission submission,
+            string userName,
+            Participant participant,
+            int totalScore)
         {
-            this.participantScores.Add(new ParticipantScore
+            participant.Scores.Add(new ParticipantScore
             {
                 ParticipantId = submission.ParticipantId.Value,
                 ProblemId = submission.ProblemId.Value,
                 SubmissionId = submission.Id,
-                ParticipantName = username,
+                ParticipantName = userName,
                 Points = submission.Points,
-                IsOfficial = isOfficial
+                IsOfficial = participant.IsOfficial
             });
+            participant.TotalScoreSnapshot = totalScore + submission.Points;
 
-            this.participantScores.SaveChanges();
+            this.participantRepository.Update(participant);
+            this.participantRepository.SaveChanges();
         }
 
         public void UpdateBySubmissionAndPoints(
             ParticipantScore participantScore,
             int? submissionId,
-            int submissionPoints)
+            int submissionPoints,
+            Participant participant,
+            int totalScore)
         {
+            participant.TotalScoreSnapshot = totalScore + submissionPoints;
```

Comment:
Make sure the Snapshot is updated properly for Retests by:
1. Single submission
2. Problem
3. Contest

Reply:
Done